qmail-local.c

```diff
@@ -80,35 +82,55 @@ void maildir_child(dir)
 char *dir;
 {
  unsigned long pid;
- unsigned long time;
+ struct timeval time;
  char myhost[64];
  char *s;
  int loop;
  struct stat st;
  int fd;
  substdio ss;
  substdio ssout;
+ stralloc hostsa = {0};
+ size_t hostnamemaxlen = 256;
 
  sig_alarmcatch(sigalrm);
  if (chdir(dir) == -1) { if (error_temp(errno)) _exit(1); _exit(2); }
- pid = getpid();
  myhost[0] = 0;
- gethostname(myhost,sizeof(myhost));
+ pid = getpid();
+ if (!stralloc_ready(&hostsa,hostnamemaxlen)) temp_nomem();
+ if (!gethostname(hostsa.s,hostnamemaxlen)) error_temp(errno);
+ hostsa.s[hostnamemaxlen-1] = 0;
+
+ s = hostsa.s;
+ for (loop = 0; loop < str_len(hostsa.s); ++loop)
+  {
+   if (hostsa.s[loop] == '/')
+    {
+     if (!stralloc_cats(&hostname,"\\057")) temp_nomem();
+     continue;
+    }
+   if (hostsa.s[loop] == ':')
+    {
+     if (!stralloc_cats(&hostname,"\\072")) temp_nomem();
+     continue;
+    }
+   if (!stralloc_append(&hostname,s+loop)) temp_nomem();
```

Comment:
Just making sure: there is no program in qmail that tries to decode these escape codes?
It is just in case any of those are encountered?
Why not hard fail instead? It is not expected to have these characters in a host name in the first place, so it might not bother anyone if it stops in these cases.

>       <name>  ::= <let>[*[<let-or-digit-or-hyphen>]<let-or-digit>]
> https://datatracker.ietf.org/doc/html/rfc952

Although, it *may* happen: `hostname /` works just fine here.

Reply:
The encoding of / and : is specified in https://cr.yp.to/proto/maildir.html, there's no need to decode it.